RFC-0001-torch-function-for-methods.md

```diff
@@ -0,0 +1,413 @@
+Python these operators are just dunder methods of the corresponding class.
+
+### Checking for compatibility
+One can check for compatibility with supported classes in the following manner:
+
+```python
+class MyTensor:
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs):
+        HANDLED_CLASSES = (MyTensor, Tensor, ...)
+        if not issubclass(t, HANDLED_CLASSES) for t in types:
+            return NotImplemented
+        # Do further processing here.
+```
+
+### Implementing a subset of the API
+One can directly follow the following procedure to implement a subset of the
+API by using a hashmap to your own implementations of a function:
+
+```python
+_TORCH_IMPLEMENTATIONS = {}
+
+def implements(torch_function):
+    def inner(f):
+        _TORCH_IMPLEMENTATIONS[torch_function] = f
+        return f
+    return inner
+
+@implements(torch.add)
+def my_add(self, other):
+    # Implementation here
+
+class MyTensor:
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs):
+        compatible = ...
+        if not compatible:
+            return NotImplemented
+        
+        if func not in _TORCH_IMPLEMENTATIONS:
+            return NotImplemented
+
+        return _TORCH_IMPLEMENTATIONS[func](*args, **kwargs)
+```
+
+### The need for `super().__torch_function__`
+To access super, one would do the following:
+```python
+class SubTensor(torch.Tensor):
+    @classmethod
+    def __torch_function__(cls, func, types, args, kwargs):
+        # Pre-processing here
+        val = super().__torch_function__(
+            func,
+            types
+            args,
+            kwargs
+        )
+        # Post processing here
+```
+
+To make the need for `super()` to be available concrete, let's consider the
+following scenario:
+
+```python
+class SubTensor(torch.Tensor):
+    @classmethod
+    def __torch_function__(...):
+        # Pre-processing
+        ret = super().__torch_function__(
+            func,
+            types
+            args,
+            kwargs
+        )
+        # Post processing
+        return ret
+
+class SubSubTensor(SubTensor):
+    def __add__(self, other):
+        # Pre-processing
+        ret = super().__add__(other)
+        # Post-processing
+        return ret
+```
+
+In this instance, with the proposed changes, processing would follow the
+`__torch_function__` protocol. This means that control would end up in
+`SubTensor.__torch_function__`, go to `Tensor.__torch_function__` from there
+and and then come to `SubSubTensor.__add__`, from where it would go to
+`Tensor.__add__`, and then back up the stack in the reverse order. This means
+that great care needs to be taken when writing `SubTensor.__torch_function__`
+to take into account the fact that it has to handle subclass methods.
+
+In general, control flow will follow this pattern:
+
+![Control flow diagram](./RFC-0001-assets/dispatch-flow.svg)
```

Comment:
Maybe the second hand-drawn diagram from @ezyang  should be added here as well as an svg. The text above about control flow for `__add__` is hard to follow without it.

Reply:
That one's rendered invalid since essentially decorating an overridden method isn't something we do anymore.